pkg/kapis/gateway/v1alpha1/handler.go

```diff
@@ -17,30 +17,44 @@ limitations under the License.
 package v1alpha1
 
 import (
+	"context"
+	"fmt"
+
 	"github.com/emicklei/go-restful"
-	"kubesphere.io/api/gateway/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apiserver/pkg/util/flushwriter"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
 	"sigs.k8s.io/controller-runtime/pkg/cache"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	"kubesphere.io/api/gateway/v1alpha1"
+
 	"kubesphere.io/kubesphere/pkg/api"
 	"kubesphere.io/kubesphere/pkg/apiserver/query"
+	"kubesphere.io/kubesphere/pkg/informers"
 	operator "kubesphere.io/kubesphere/pkg/models/gateway"
+
 	servererr "kubesphere.io/kubesphere/pkg/server/errors"
 	"kubesphere.io/kubesphere/pkg/simple/client/gateway"
+	conversionsv1 "kubesphere.io/kubesphere/pkg/utils/conversions/core/v1"
 )
 
 type handler struct {
 	options *gateway.Options
 	gw      operator.GatewayOperator
+	factory informers.InformerFactory
 }
 
 //newHandler create an instance of the handler
-func newHandler(options *gateway.Options, cache cache.Cache, client client.Client) *handler {
+func newHandler(options *gateway.Options, cache cache.Cache, client client.Client, factory informers.InformerFactory, k8sClient kubernetes.Interface) *handler {
+	conversionsv1.RegisterConversions(scheme.Scheme)
 	// Do not register Gateway scheme globally. Which will cause conflict in ks-controller-manager.
 	v1alpha1.AddToScheme(client.Scheme())
```

Comment:
Why not register the scheme in `pkg/apis/apis.go` 

https://github.com/kubesphere/kubesphere/blob/master/cmd/ks-apiserver/app/options/options.go#L227

Reply:
We introduced an SDK in `ks-controller-manager` called `Helm Operator plugin`. Which will register an unstructured scheme `Gateway` firstly. If we register the typed `Gateway` scheme in `pkg/apis/apis.go`, the controller-runtime client will throw an exception.